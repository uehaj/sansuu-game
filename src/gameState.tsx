import React, { useState } from 'react';

type GamePhase = 'initial' | 'splash' | 'ready' | 'running';

export interface GameState {
  gamePhase: GamePhase;
  gameRound: number;
}

export type GameUpdator = {
  setGamePhase: (phase: GamePhase) => void;
  setGameRound: (round: number) => void;
};

export function useGameState(): [GameState, GameUpdator] {
  const [gameState, setGameState] = useState<GameState>({
    gamePhase: 'initial',
    gameRound: 0,
  });
  return [
    gameState,
    {
      setGamePhase: (newPhase: GamePhase) =>
        setGameState((prevState: GameState) => ({
          ...prevState,
          gamePhase: newPhase,
        })),
      setGameRound: (newRound: number) =>
        setGameState(prevState => ({
          ...prevState,
          gameRound: newRound,
        })),
    },
  ];
}
