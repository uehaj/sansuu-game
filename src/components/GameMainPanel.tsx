import React, { useCallback } from 'react';
import CssBaseline from '@material-ui/core/CssBaseline';

import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import MenuIcon from '@material-ui/icons/Menu';
import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';

import NumPad from './NumPad';
import { useGameState } from '../gameState';

const styles = {
  button: {
    margin: '1rem',
    fontSize: '2rem',
  },
};

type Props = {
  classes: any;
};

const gameRun = () => useCallback(k => console.log(k), []);

function GameMainPanel({ classes }: Props) {
  const [gameState, gameUpdator] = useGameState();
  const gameStart = useCallback(() => {
    gameUpdator.setGamePhase('running');
  }, []);

  return (
    <Grid container>
      <Grid item xs={1} />
      {(state => {
        console.log(state);
        if (state === 'initial') {
          gameUpdator.setGamePhase('splash');
          return <div />;
        } else if (state === 'splash') {
          setTimeout(() => {
            gameUpdator.setGamePhase('ready');
          }, 1000);
          return (
            <Typography align={'center'} variant={'h1'}>
              算数ゲーム
            </Typography>
          );
        } else if (state === 'ready') {
          return (
            <Grid item xs={10}>
              <Button
                variant="contained"
                className={classes.button}
                color={'secondary'}
                onClick={gameStart}>
                START
              </Button>
            </Grid>
          );
        } else if (state === 'running') {
          return (
            <Grid item xs={10}>
              <NumPad onChange={gameRun} />
            </Grid>
          );
        } else {
          return <div>Error</div>;
        }
      })(gameState.gamePhase)}
    </Grid>
  );
}

export default withStyles(styles)(GameMainPanel);
