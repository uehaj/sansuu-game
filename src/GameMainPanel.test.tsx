import React, { useState } from 'react';
import ReactDOM from 'react-dom';
import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
} from 'react-testing-library';
import 'jest-dom/extend-expect';

import { useGameState } from './gameState';
import GameMainPanel from './components/GameMainPanel';

afterEach(cleanup);

test('GameMainPanelを読み込んでSTARTボタンを押すとゲーム画面が表示される', async () => {
  const { getByTestId, container } = render(<GameMainPanel />);
  const startButton = await waitForElement(() => getByTestId('start'));
  fireEvent.click(startButton);
  expect(getByTestId('text')).toHaveTextContent('aaa');
  expect(container.firstChild).toMatchSnapshot();
});
