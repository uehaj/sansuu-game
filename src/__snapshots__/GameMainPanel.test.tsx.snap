// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameMainPanelを読み込んでSTARTボタンを押すとゲーム画面が表示される 1`] = `
<div
  class="MuiGrid-container-2"
>
  <div
    class="MuiGrid-item-3 MuiGrid-grid-xs-1-31"
  />
  <div
    class="MuiGrid-item-3 MuiGrid-grid-xs-10-40"
  >
    <div
      class="MuiPaper-root-174 MuiPaper-elevation2-178 MuiPaper-rounded-175"
      style="margin: 1rem;"
    >
      <div
        class="MuiGrid-container-2"
      >
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-12-42 NumPad-textArea-172"
          data-testid="text"
        >
          aaa
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-12-42 NumPad-textArea-172"
        >
          <div
            class="MuiPaper-root-174 MuiPaper-elevation2-178 MuiPaper-rounded-175"
          >
            <p
              class="MuiTypography-root-201 MuiTypography-body1-210 MuiTypography-alignCenter-224 NumPad-inputArea-173"
            >
              0
            </p>
          </div>
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-4-34"
          style="text-align: center;"
        >
          <button
            class="MuiButtonBase-root-161 MuiButton-root-135 MuiButton-contained-146 MuiButton-raised-149 NumPad-numButton-171"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-136"
            >
              0
            </span>
            <span
              class="MuiTouchRipple-root-237"
            />
          </button>
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-4-34"
          style="text-align: center;"
        >
          <button
            class="MuiButtonBase-root-161 MuiButton-root-135 MuiButton-contained-146 MuiButton-raised-149 NumPad-numButton-171"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-136"
            >
              1
            </span>
            <span
              class="MuiTouchRipple-root-237"
            />
          </button>
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-4-34"
          style="text-align: center;"
        >
          <button
            class="MuiButtonBase-root-161 MuiButton-root-135 MuiButton-contained-146 MuiButton-raised-149 NumPad-numButton-171"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-136"
            >
              2
            </span>
            <span
              class="MuiTouchRipple-root-237"
            />
          </button>
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-4-34"
          style="text-align: center;"
        >
          <button
            class="MuiButtonBase-root-161 MuiButton-root-135 MuiButton-contained-146 MuiButton-raised-149 NumPad-numButton-171"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-136"
            >
              3
            </span>
            <span
              class="MuiTouchRipple-root-237"
            />
          </button>
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-4-34"
          style="text-align: center;"
        >
          <button
            class="MuiButtonBase-root-161 MuiButton-root-135 MuiButton-contained-146 MuiButton-raised-149 NumPad-numButton-171"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-136"
            >
              4
            </span>
            <span
              class="MuiTouchRipple-root-237"
            />
          </button>
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-4-34"
          style="text-align: center;"
        >
          <button
            class="MuiButtonBase-root-161 MuiButton-root-135 MuiButton-contained-146 MuiButton-raised-149 NumPad-numButton-171"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-136"
            >
              5
            </span>
            <span
              class="MuiTouchRipple-root-237"
            />
          </button>
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-4-34"
          style="text-align: center;"
        >
          <button
            class="MuiButtonBase-root-161 MuiButton-root-135 MuiButton-contained-146 MuiButton-raised-149 NumPad-numButton-171"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-136"
            >
              6
            </span>
            <span
              class="MuiTouchRipple-root-237"
            />
          </button>
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-4-34"
          style="text-align: center;"
        >
          <button
            class="MuiButtonBase-root-161 MuiButton-root-135 MuiButton-contained-146 MuiButton-raised-149 NumPad-numButton-171"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-136"
            >
              7
            </span>
            <span
              class="MuiTouchRipple-root-237"
            />
          </button>
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-4-34"
          style="text-align: center;"
        >
          <button
            class="MuiButtonBase-root-161 MuiButton-root-135 MuiButton-contained-146 MuiButton-raised-149 NumPad-numButton-171"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-136"
            >
              8
            </span>
            <span
              class="MuiTouchRipple-root-237"
            />
          </button>
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-4-34"
          style="text-align: center;"
        >
          <button
            class="MuiButtonBase-root-161 MuiButton-root-135 MuiButton-contained-146 MuiButton-raised-149 NumPad-numButton-171"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-136"
            >
              9
            </span>
            <span
              class="MuiTouchRipple-root-237"
            />
          </button>
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-4-34"
          style="text-align: center;"
        >
          <button
            class="MuiButtonBase-root-161 MuiButton-root-135 MuiButton-contained-146 MuiButton-raised-149 NumPad-numButton-171"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-136"
            >
              CLR
            </span>
            <span
              class="MuiTouchRipple-root-237"
            />
          </button>
        </div>
        <div
          class="MuiGrid-item-3 MuiGrid-grid-xs-4-34"
          style="text-align: center;"
        >
          <button
            class="MuiButtonBase-root-161 MuiButton-root-135 MuiButton-contained-146 MuiButton-raised-149 NumPad-numButton-171"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label-136"
            >
              OK
            </span>
            <span
              class="MuiTouchRipple-root-237"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
